### PR TITLE
Add Toitware's esp-idf fork as submodule

### PR DESCRIPTION
**Pull request by mikini**
* add .gitmodules pointing to branch patch-head-4.3-3
* register current head of branch patch-head-4.3-3 as eps-idf's index

Note: currently there is a reference problem on the fork, see #45.

### CODE REVIEW

**Comment by mikini** [comment]
Had the same problem as #45 earlier doing this. But seems to work fine now (this is also on another machine (Ubuntu 20.04, failing one was 16.04);
```
:~$ git clone https://github.com/mikini/toit toit_mikini
Cloning into 'toit_mikini'...
remote: Enumerating objects: 1125, done.
remote: Counting objects: 100% (1125/1125), done.
remote: Compressing objects: 100% (891/891), done.
remote: Total 1125 (delta 314), reused 1011 (delta 210), pack-reused 0
Receiving objects: 100% (1125/1125), 4.91 MiB | 2.68 MiB/s, done.
Resolving deltas: 100% (314/314), done.
:~$ cd toit_mikini/
:~/toit_mikini (master)$ git co esp-idf-as-submodule 
Branch 'esp-idf-as-submodule' set up to track remote branch 'esp-idf-as-submodule' from 'origin'.
Switched to a new branch 'esp-idf-as-submodule'
:~/toit_mikini (esp-idf-as-submodule)$ git submodule update --init --recursive
Submodule 'esp-idf' (https://github.com/toitware/esp-idf.git) registered for path 'esp-idf'
Cloning into '/home/miki/toit_mikini/esp-idf'...
Submodule path 'esp-idf': checked out '54e724e0506259dae79ae457e35251447ed20a1e'
Submodule 'components/asio/asio' (https://github.com/espressif/asio.git) registered for path 'esp-idf/components/asio/asio'
Submodule 'components/bootloader/subproject/components/micro-ecc/micro-ecc' (https://github.com/kmackay/micro-ecc.git) registered for path 'esp-idf/components/bootloader/subproject/components/micro-ecc/micro-ecc'
Submodule 'components/bt/controller/lib_esp32' (https://github.com/espressif/esp32-bt-lib.git) registered for path 'esp-idf/components/bt/controller/lib_esp32'
Submodule 'components/bt/controller/lib_esp32c3_family' (https://github.com/espressif/esp32c3-bt-lib.git) registered for path 'esp-idf/components/bt/controller/lib_esp32c3_family'
Submodule 'components/bt/host/nimble/nimble' (https://github.com/espressif/esp-nimble.git) registered for path 'esp-idf/components/bt/host/nimble/nimble'
Submodule 'components/cbor/tinycbor' (https://github.com/intel/tinycbor.git) registered for path 'esp-idf/components/cbor/tinycbor'
Submodule 'components/cmock/CMock' (https://github.com/ThrowTheSwitch/CMock.git) registered for path 'esp-idf/components/cmock/CMock'
Submodule 'components/coap/libcoap' (https://github.com/obgm/libcoap.git) registered for path 'esp-idf/components/coap/libcoap'
Submodule 'components/esp_wifi/lib' (https://github.com/espressif/esp32-wifi-lib.git) registered for path 'esp-idf/components/esp_wifi/lib'
Submodule 'components/esptool_py/esptool' (https://github.com/espressif/esptool.git) registered for path 'esp-idf/components/esptool_py/esptool'
Submodule 'components/expat/expat' (https://github.com/libexpat/libexpat.git) registered for path 'esp-idf/components/expat/expat'
Submodule 'components/json/cJSON' (https://github.com/DaveGamble/cJSON.git) registered for path 'esp-idf/components/json/cJSON'
Submodule 'components/libsodium/libsodium' (https://github.com/jedisct1/libsodium.git) registered for path 'esp-idf/components/libsodium/libsodium'
Submodule 'components/lwip/lwip' (https://github.com/espressif/esp-lwip.git) registered for path 'esp-idf/components/lwip/lwip'
Submodule 'components/mbedtls/mbedtls' (https://github.com/espressif/mbedtls.git) registered for path 'esp-idf/components/mbedtls/mbedtls'
Submodule 'components/mqtt/esp-mqtt' (https://github.com/espressif/esp-mqtt.git) registered for path 'esp-idf/components/mqtt/esp-mqtt'
Submodule 'components/nghttp/nghttp2' (https://github.com/nghttp2/nghttp2.git) registered for path 'esp-idf/components/nghttp/nghttp2'
Submodule 'components/protobuf-c/protobuf-c' (https://github.com/protobuf-c/protobuf-c.git) registered for path 'esp-idf/components/protobuf-c/protobuf-c'
Submodule 'components/spiffs/spiffs' (https://github.com/pellepl/spiffs.git) registered for path 'esp-idf/components/spiffs/spiffs'
Submodule 'components/tinyusb/tinyusb' (https://github.com/espressif/tinyusb.git) registered for path 'esp-idf/components/tinyusb/tinyusb'
Submodule 'components/unity/unity' (https://github.com/ThrowTheSwitch/Unity.git) registered for path 'esp-idf/components/unity/unity'
Submodule 'examples/build_system/cmake/import_lib/main/lib/tinyxml2' (https://github.com/leethomason/tinyxml2.git) registered for path 'esp-idf/examples/build_system/cmake/import_lib/main/lib/tinyxml2'
Submodule 'examples/peripherals/secure_element/atecc608_ecdsa/components/esp-cryptoauthlib' (https://github.com/espressif/esp-cryptoauthlib.git) registered for path 'esp-idf/examples/peripherals/secure_element/atecc608_ecdsa/components/esp-cryptoauthlib'
Cloning into '/home/miki/toit_mikini/esp-idf/components/asio/asio'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/bootloader/subproject/components/micro-ecc/micro-ecc'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/bt/controller/lib_esp32'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/bt/controller/lib_esp32c3_family'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/bt/host/nimble/nimble'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/cbor/tinycbor'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/cmock/CMock'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/coap/libcoap'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/esp_wifi/lib'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/esptool_py/esptool'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/expat/expat'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/json/cJSON'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/libsodium/libsodium'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/lwip/lwip'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/mbedtls/mbedtls'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/mqtt/esp-mqtt'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/nghttp/nghttp2'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/protobuf-c/protobuf-c'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/spiffs/spiffs'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/tinyusb/tinyusb'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/unity/unity'...
Cloning into '/home/miki/toit_mikini/esp-idf/examples/build_system/cmake/import_lib/main/lib/tinyxml2'...
Cloning into '/home/miki/toit_mikini/esp-idf/examples/peripherals/secure_element/atecc608_ecdsa/components/esp-cryptoauthlib'...
Submodule path 'esp-idf/components/asio/asio': checked out 'f31694c9f1746ba189a4bcae2e34db15135ddb22'
Submodule path 'esp-idf/components/bootloader/subproject/components/micro-ecc/micro-ecc': checked out 'd037ec89546fad14b5c4d5456c2e23a71e554966'
Submodule path 'esp-idf/components/bt/controller/lib_esp32': checked out 'fb49791b7c1a8a35f06e68124c90022667b4cff1'
Submodule path 'esp-idf/components/bt/controller/lib_esp32c3_family': checked out '12f00c45ce9c8cf9a9b2e607b4954f12d4191ffb'
Submodule path 'esp-idf/components/bt/host/nimble/nimble': checked out 'aef55bbf636ed580d4d6408a5c2e75d1f70a875e'
Submodule path 'esp-idf/components/cbor/tinycbor': checked out '085ca40781f7c39febe6d14fb7e5cba342e1804b'
Submodule path 'esp-idf/components/cmock/CMock': checked out 'eeecc49ce8af123cf8ad40efdb9673e37b56230f'
Submodule 'vendor/c_exception' (https://github.com/throwtheswitch/cexception.git) registered for path 'esp-idf/components/cmock/CMock/vendor/c_exception'
Submodule 'vendor/unity' (https://github.com/throwtheswitch/unity.git) registered for path 'esp-idf/components/cmock/CMock/vendor/unity'
Cloning into '/home/miki/toit_mikini/esp-idf/components/cmock/CMock/vendor/c_exception'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/cmock/CMock/vendor/unity'...
Submodule path 'esp-idf/components/cmock/CMock/vendor/c_exception': checked out '71b47be7c950f1bf5f7e5303779fa99a16224bb6'
Submodule path 'esp-idf/components/cmock/CMock/vendor/unity': checked out 'cf949f45ca6d172a177b00da21310607b97bc7a7'
Submodule path 'esp-idf/components/coap/libcoap': checked out '98954eb30a2e728e172a6cd29430ae5bc999b585'
Submodule 'ext/tinydtls' (https://github.com/eclipse/tinydtls.git) registered for path 'esp-idf/components/coap/libcoap/ext/tinydtls'
Cloning into '/home/miki/toit_mikini/esp-idf/components/coap/libcoap/ext/tinydtls'...
Submodule path 'esp-idf/components/coap/libcoap/ext/tinydtls': checked out '7f8c86e501e690301630029fa9bae22424adf618'
Submodule path 'esp-idf/components/esp_wifi/lib': checked out 'e6945e61f7c63545a77b0575c3770a85b4de948e'
Submodule path 'esp-idf/components/esptool_py/esptool': checked out '4698b396730b23fb4aab023c5fb1744db957fc4c'
Submodule path 'esp-idf/components/expat/expat': checked out '968b8cc46dbee47b83318d5f31a8e7907199614b'
Submodule path 'esp-idf/components/json/cJSON': checked out 'd348621ca93571343a56862df7de4ff3bc9b5667'
Submodule path 'esp-idf/components/libsodium/libsodium': checked out '4f5e89fa84ce1d178a6765b8b46f2b6f91216677'
remote: Enumerating objects: 62, done.
remote: Counting objects: 100% (62/62), done.
remote: Compressing objects: 100% (40/40), done.
remote: Total 62 (delta 22), reused 60 (delta 22), pack-reused 0
Unpacking objects: 100% (62/62), 54.42 KiB | 592.00 KiB/s, done.
From https://github.com/espressif/esp-lwip
 * branch              02638065e72ea689eb0d3509b1960b31c515616e -> FETCH_HEAD
Submodule path 'esp-idf/components/lwip/lwip': checked out '02638065e72ea689eb0d3509b1960b31c515616e'
Submodule path 'esp-idf/components/mbedtls/mbedtls': checked out '6465247f67167518b8813ae2faaf422704e4b1a3'
Submodule path 'esp-idf/components/mqtt/esp-mqtt': checked out 'f10321a53b53a146ee299cfecc320b89c0cf6611'
Submodule path 'esp-idf/components/nghttp/nghttp2': checked out '8f7b008b158e12de0e58247afd170f127dbb6456'
Submodule 'third-party/mruby' (https://github.com/mruby/mruby) registered for path 'esp-idf/components/nghttp/nghttp2/third-party/mruby'
Submodule 'third-party/neverbleed' (https://github.com/tatsuhiro-t/neverbleed.git) registered for path 'esp-idf/components/nghttp/nghttp2/third-party/neverbleed'
Cloning into '/home/miki/toit_mikini/esp-idf/components/nghttp/nghttp2/third-party/mruby'...
Cloning into '/home/miki/toit_mikini/esp-idf/components/nghttp/nghttp2/third-party/neverbleed'...
Submodule path 'esp-idf/components/nghttp/nghttp2/third-party/mruby': checked out '7c91efc1ffda769a5f1a872c646c82b00698f1b8'
Submodule path 'esp-idf/components/nghttp/nghttp2/third-party/neverbleed': checked out 'b967ca054f48a36f82d8fcdd32e54ec5144f2751'
Submodule path 'esp-idf/components/protobuf-c/protobuf-c': checked out 'dac1a65feac4ad72f612aab99f487056fbcf5c1a'
Submodule path 'esp-idf/components/spiffs/spiffs': checked out 'f5e26c4e933189593a71c6b82cda381a7b21e41c'
Submodule path 'esp-idf/components/tinyusb/tinyusb': checked out '334e95fac52a607150157ae5199a19e11f843982'
Submodule path 'esp-idf/components/unity/unity': checked out '7d2bf62b7e6afaf38153041a9d53c21aeeca9a25'
Submodule path 'esp-idf/examples/build_system/cmake/import_lib/main/lib/tinyxml2': checked out '7e8e249990ec491ec15990cf95b6d871a66cf64a'
Submodule path 'esp-idf/examples/peripherals/secure_element/atecc608_ecdsa/components/esp-cryptoauthlib': checked out 'bb672b0437485fc7420add178299631692b15ac3'
:~/toit_mikini (esp-idf-as-submodule)$ export IDF_PATH=$PWD/esp-idf/
:~/toit_mikini (esp-idf-as-submodule)$ make tools
GO111MODULE=on GOBIN=/home/miki/toit_mikini/build go get github.com/toitlang/tpkg/cmd/toitpkg@"v0.0.0-20211126161923-c00da039da00"
go: finding github.com/toitlang v0.0.0-20211126161923-c00da039da00
go: finding github.com/toitlang/tpkg/cmd v0.0.0-20211126161923-c00da039da00
go: finding github.com v0.0.0-20211126161923-c00da039da00
go: finding github.com/toitlang/tpkg/cmd/toitpkg v0.0.0-20211126161923-c00da039da00
cd tools/toitlsp; CGO_ENABLED=1 GODEBUG=netdns=go go build  -ldflags " -X main.date=2021-12-02T16:37:37Z" -tags 'netgo osusergo' -o ../../build/toitlsp .
mkdir -p build/host/
(cd build/host && cmake ../.. -G Ninja -DCMAKE_BUILD_TYPE=Release)
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PythonInterp: /usr/bin/python (found version "2.7.18") 
-- Found Perl: /usr/bin/perl (found version "5.30.0") 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Configuring done
-- Generating done
-- Build files have been written to: /home/miki/toit_mikini/build/host
(cd build/host && ninja build_toitvm)
[322/322] Generating ../bin/toitvm_boot.snapshot
(cd build/host && ninja build_toitvm)
[1/1] Generating ../bin/toitvm_boot.snapshot
:~/toit_mikini (esp-idf-as-submodule)$ build/host/bin/toitvm examples/hello.toit
Hello, World!
:~/toit_mikini (esp-idf-as-submodule)$ 

```

**Comment by kasperl** [comment]
Given the requirement to deal with git submodules anyway (for the esp-idf dependencies), I think this makes sense. I'd suggest pulling in the `esp-idf` directory in `third_party/esp-idf` so we have a common place for more good stuff we will come to rely on in the future.

File: README.md [hunk 1]
```diff
@@ -75,12 +75,11 @@ We recommend you use Toitware's [ESP-IDF fork](https://github.com/toitware/esp-i
 * Allocation-fixes for UART, etc.
 * LWIP fixes.
 
+This repository has been added as as submodule reference, so doing a recursive init & update will establish everything nedded:
```
Comment (kasperl) [inline README.md:78]: We should update the instructions for how to set IDF_PATH to indicate where this is found - and indicate that this is where you would choose a different `esp-idf` version if you want to tinker with it.

```
export IDF_PATH=`pwd`/third_party/esp-idf
```

Comment (kasperl) [reply README.md:78]: We could also let the Makefile give these instructions when it finds that `IDF_PATH` isn't set.

Comment (mikini) [reply README.md]: Added as suggested, plus some copyediting.

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,4 @@
+[submodule "esp-idf"]
+	path = third_party/esp-idf/esp-idf
```
Comment (kasperl) [inline .gitmodules:2]: Let's make this `third_party/esp-idf` and avoid the double nesting.

Comment (mikini) [reply .gitmodules]: Hmm, toyed around with git rm's automatic .gitmodule update feature. It must have fooled me somehow. Fixed.

**Comment by kasperl** [comment]
I continued the work on this here: https://github.com/toitlang/toit/pull/82. I found that I had to update the added third_party/esp-idf to get it to point to the correct commit from the patch-head-4.3-3 branch.

**Comment by kasperl** [comment]
Merged in via 186556eb46f13fa4befeb073dea79de6663ce67e.

**Pull request closed without merging**